Storage/feature/clientside encryption (blob)

jaschrep-msft opened the pull request:
Blobs now support client-side encryption for block blobs in Azure.Storage.Blobs.Cryptography.

jaschrep-msft committed:
Initial helper class port.

jaschrep-msft committed:
Merge branch 'master' of https://github.com/Azure/azure-sdk-for-net into storage/feature/clientside-encryption

jaschrep-msft committed:
Checkpoint

jaschrep-msft committed:
Checkpoint

jaschrep-msft committed:
Checkpoint. 90% groudwork laid.

jaschrep-msft committed:
Checkpoint

jaschrep-msft committed:
Changed Inheritance structure

Now inherit from BlobClient instead of BlobBaseClient

jaschrep-msft committed:
Removed unnecessary class

jaschrep-msft committed:
Merge branch 'master' of https://github.com/Azure/azure-sdk-for-net into storage/feature/clientside-encryption

jaschrep-msft committed:
Checkpoint

jaschrep-msft committed:
Checkpoint

jaschrep-msft committed:
Merge branch 'master' of https://github.com/Azure/azure-sdk-for-net into storage/feature/clientside-encryption

jaschrep-msft committed:
Rolling Buffer started

Checkpoint

jaschrep-msft committed:
RollingBufferStream works

jaschrep-msft committed:
Rolling Buffer fixed

jaschrep-msft committed:
Encryption working

jaschrep-msft committed:
Merge branch 'master' of https://github.com/Azure/azure-sdk-for-net into storage/feature/clientside-encryption

jaschrep-msft committed:
Undo temp hack

jaschrep-msft commented:
Needs some organizational cleanup. Will address alongside PR comments.

pakrym commented on sdk/core/Azure.Core/src/Shared/NullableAttributes.cs:
```diff
@@ -15,7 +15,8 @@ namespace System.Diagnostics.CodeAnalysis
 #else
     public
 #endif
-        sealed class AllowNullAttribute : Attribute { }
```
Revert

jaschrep-msft replied:
Visual studio must have autoformatted at some point. Will revert.

amnguye commented on sdk/storage/Azure.Storage.Blobs.Cryptography/src/ClientSideDecryptionPolicy.cs:
```diff
@@ -0,0 +1,393 @@
+            else
+            {
+                key = this.KeyResolver?.Resolve(encryptionData.WrappedContentKey.KeyId) ?? throw new ArgumentException(
+                    "Key mismatch. The resolver could not match the key specified by the service.");
```
Make a new error in BlobErrors.cs for this

seanmcc-msft replied:
Same.  Or maybe make an EncryptionErrors class, similar to Azure.Storage.Common.Errors?

seanmcc-msft commented on sdk/storage/Azure.Storage.Blobs.Cryptography/src/ClientSideDecryptionPolicy.cs:
```diff
@@ -0,0 +1,393 @@
+
+        private static Metadata ExtractMetadata(ResponseHeaders headers)
+        {
+            const string metadataPrefix = "x-ms-meta-";
```
Consider moving to EncryptionConstants

amnguye replied:
make a constant for this

seanmcc-msft commented on sdk/storage/Azure.Storage.Blobs.Cryptography/src/ClientSideDecryptionPolicy.cs:
```diff
@@ -0,0 +1,393 @@
+            EncryptionData encryptionData;
+            if (!metadata.TryGetValue(EncryptionConstants.ENCRYPTION_DATA_KEY, out string encryptedDataString))
+            {
+                throw new InvalidOperationException("Encryption data does not exist.");
```
Consider moving to EncryptionContants

jaschrep-msft replied:
We are now downloading normally if data isn't encrypted. Saves customers a call if they mess up.

amnguye commented on sdk/storage/Azure.Storage.Blobs.Cryptography/src/ClientSideDecryptionPolicy.cs:
```diff
@@ -0,0 +1,393 @@
+                        return new RollingBufferStream(new CryptoStream(contentStream, aesProvider.CreateDecryptor(), CryptoStreamMode.Read), 10 * Constants.MB);
+                    }
+                default:
+                    throw new ArgumentException(
```
Make a new error in BlobErrors.cs for this

seanmcc-msft replied:
Same

amnguye commented on sdk/storage/Azure.Storage.Blobs.Cryptography/src/ClientSideDecryptionPolicy.cs:
```diff
@@ -0,0 +1,393 @@
+            else
+            {
+                key = await (this.KeyResolver?.ResolveAsync(encryptionData.WrappedContentKey.KeyId) ??
+                    throw new ArgumentException(
```
Make a new error in BlobErrors.cs for this

seanmcc-msft replied:
Same as above.  Maybe make an EncryptionErrors class?

amnguye commented on sdk/storage/Azure.Storage.Blobs.Cryptography/src/ClientSideDecryptionPolicy.cs:
```diff
@@ -0,0 +1,393 @@
+        {
+            if (this.LocalKey == default && this.KeyResolver == default)
+            {
+                throw new InvalidOperationException("Key and KeyResolver cannot both be null");
```
Make a new error in BlobErrors.cs for this

seanmcc-msft commented on sdk/storage/Azure.Storage.Blobs.Cryptography/src/EncryptedBlobClient.cs:
```diff
@@ -0,0 +1,330 @@
+
+                var encryptionData = new EncryptionData()
+                {
+                    EncryptionMode = "FullBlob",
```
Consider moving to EncryptionConstants

amnguye replied:
constant for this?

seanmcc-msft commented on sdk/storage/Azure.Storage.Blobs.Cryptography/src/EncryptedBlobClient.cs:
```diff
@@ -0,0 +1,330 @@
+
+                return (
+                    new RollingBufferStream(
+                        new CryptoStream(plaintext, aesProvider.CreateEncryptor(), CryptoStreamMode.Read), 10 * Constants.MB,
```
Consider moving to Encryption constants, 10 * Constants.MB as something along the lines of DefaultLengthLimitingStreamSize, and 16 as BlockSize.

seanmcc-msft commented on sdk/storage/Azure.Storage.Blobs.Cryptography/src/LengthLimitingStream.cs:
```diff
@@ -0,0 +1,81 @@
+            _wrappedStream.Dispose();
+        }
+
+        private int Clamp(int count)
```
I think it would be good to add XML documentation for this method.

seanmcc-msft commented on sdk/storage/Azure.Storage.Blobs.Cryptography/src/Models/EncryptedBlobRange.cs:
```diff
@@ -0,0 +1,72 @@
+    /// </summary>
+    internal struct EncryptedBlobRange
+    {
+        public const int ENCRYPTION_BLOCK_SIZE = 16;
```
Move to EncryptionConstants.

amnguye replied:
This is already in EncryptionConstants, you probably dont need this here.

seanmcc-msft commented on sdk/storage/Azure.Storage.Blobs/src/BlobBaseClient.cs:
```diff
@@ -1227,8 +1227,18 @@ Task<Response<BlobProperties>> GetPropertiesAsync(bool async, CancellationToken
+                if (async)
+                {
+                    await response.Value.Content.CopyToAsync(
+                        destination, 81920 /* default value */, ct).ConfigureAwait(false);
```
Consider moving 81920 to Azure.Storage.Common.Constants.

jaschrep-msft replied:
I only moved it. I do not know what it means and so don't know what to call it. @tg-msft?

tg-msft replied:
Yeah...  this was a fun one.  Different versions of the .NET framework use different default buffer sizes for `CopyToAsync` which caused a couple of tiny deltas in recorded tests.  We worked around it by fixing the default buffer size to the value in .NET Core.  It's fine to make this a constant if we want.

seanmcc-msft commented on sdk/storage/Azure.Storage.Blobs/src/RollingBufferStream.cs:
```diff
@@ -0,0 +1,210 @@
+        private long? _expectedLength;
+        public override long Length => _expectedLength ?? _underlyingStreamBytesRead;
+
+        public override long Position
```
I think it would be good to add more inline comments in this method.

seanmcc-msft commented on sdk/storage/Azure.Storage.Blobs/src/RollingBufferStream.cs:
```diff
@@ -0,0 +1,210 @@
+             */
+        }
+
+        public override int Read(byte[] buffer, int offset, int count)
```
Well documented.

amnguye commented on sdk/storage/Azure.Storage.Blobs.Cryptography/src/ClientSideDecryptionPolicy.cs:
```diff
@@ -0,0 +1,393 @@
+
+            // did we request the last block?
+            if (long.Parse(tokens[3], System.Globalization.CultureInfo.InvariantCulture) -
+                long.Parse(tokens[2], System.Globalization.CultureInfo.InvariantCulture) < 16)
```
What is 16? Remove magic values/make a constant for this

amnguye commented on sdk/storage/Azure.Storage.Blobs.Cryptography/src/ClientSideDecryptionPolicy.cs:
```diff
@@ -0,0 +1,393 @@
+                read = IV.Length;
+            }
+
+            var plaintext = WrapStream(ciphertext, (await GetContentEncryptionKeyAsync(encryptionData).ConfigureAwait(false)).ToArray(), encryptionData, IV, noPadding);
```
new line for each parameter

amnguye commented on sdk/storage/Azure.Storage.Blobs.Cryptography/src/ClientSideDecryptionPolicy.cs:
```diff
@@ -0,0 +1,393 @@
+            // understands
+            // and is able to decrypt.
+            if (EncryptionConstants.ENCRYPTION_PROTOCOL_V1 != encryptionData.EncryptionAgent.Protocol) {
+                throw new ArgumentException(
```
Make a new error in BlobErrors.cs for this

amnguye commented on sdk/storage/Azure.Storage.Blobs.Cryptography/src/ClientSideDecryptionPolicy.cs:
```diff
@@ -0,0 +1,393 @@
+                    return new HttpRange(firstByte, lastByteInclusive - firstByte + 1);
+
+                default:
+                    throw new ArgumentException("Could not parse the serialized range.", nameof(serializedRange));
```
Make a new error in BlobErrors.cs for this

amnguye commented on sdk/storage/Azure.Storage.Blobs.Cryptography/src/EncryptedBlobClient.cs:
```diff
@@ -0,0 +1,330 @@
+            IKeyEncryptionKeyResolver keyResolver = default,
+            string encryptionKeyWrapAlgorithm = default,
+            BlobClientOptions options = default)
+            : base(connectionString, containerName, blobName, FluentAddPolicy(options, new ClientSideDecryptionPolicy(keyResolver, keyEncryptionKey)))
```
NIT - Maybe consider making a new line for each parameter for this? If not I'm okay with it

amnguye commented on sdk/storage/Azure.Storage.Blobs.Cryptography/src/EncryptedBlobClient.cs:
```diff
@@ -0,0 +1,330 @@
+        ///// policies for authentication, retries, etc., that are applied to
+        ///// every request.
+        ///// </param>
+        //public EncryptedBlockBlobClient(
```
will this be used later?

jaschrep-msft replied:
I think I knocked this one out because I found it unlikely that we'd need to support clientside encryption on a public access blob.

amnguye commented on sdk/storage/Azure.Storage.Blobs.Cryptography/src/EncryptedBlobClient.cs:
```diff
@@ -0,0 +1,330 @@
+                return (
+                    new RollingBufferStream(
+                        new CryptoStream(plaintext, aesProvider.CreateEncryptor(), CryptoStreamMode.Read), 10 * Constants.MB,
+                        plaintext.Length + (16 - plaintext.Length % 16)),
```
Is there a reason why its 10? Maybe should also add it to constants

amnguye reviewed (approved):
Looks good!

jaschrep-msft committed:
Docs/Constants cleanup

tg-msft reviewed (commented):
(I'll start adding comments individually instead of as one big review)

tg-msft commented on sdk/storage/Azure.Storage.Blobs.Cryptography/src/Azure.Storage.Blobs.Cryptography.csproj:
```diff
@@ -17,6 +17,7 @@
       Microsoft Azure Storage REST API Reference - https://docs.microsoft.com/en-us/rest/api/storageservices/
       REST API Reference for Blob Service - https://docs.microsoft.com/en-us/rest/api/storageservices/blob-service-rest-api
     </Description>
+    <RootNamespace>Azure.Storage.Blobs.Specialized.Cryptography</RootNamespace>
```
Is there a reason we're not putting everything in `.Specialized`?

jaschrep-msft replied:
Response in previous comment on this topic.

tg-msft commented on sdk/storage/Azure.Storage.Blobs.Cryptography/src/ClientSideDecryptionPolicy.cs:
```diff
@@ -0,0 +1,412 @@
+
+            ProcessNext(message, pipeline);
+
+            if (message.Request.Method != RequestMethod.Head &&
```
Should this just explicitly check for `GET`?

tg-msft commented on sdk/storage/Azure.Storage.Blobs.Cryptography/src/ClientSideDecryptionPolicy.cs:
```diff
@@ -0,0 +1,412 @@
+            Metadata metadata = new Dictionary<string, string>(StringComparer.OrdinalIgnoreCase);
+            foreach (HttpHeader header in headers)
+            {
+                if (header.Name.StartsWith(Constants.HeaderNames.MetadataPrefix, StringComparison.InvariantCulture))
```
Do you want `OrdinalIgnoreCase` as well here since this is also case insensitive?

jaschrep-msft replied:
Interesting. The service documentation doesn't actually explain how the x-ms-meta- prefix interacts with the rest of the identifier being "case insensitive but preserving." We'll add it just in case, but I don't know if that's possible or not.

tg-msft commented on sdk/storage/Azure.Storage.Blobs.Cryptography/src/ClientSideDecryptionPolicy.cs:
```diff
@@ -0,0 +1,412 @@
+        /// <returns>The relevant metadata.</returns>
+        private static EncryptionData? GetAndValidateEncryptionData(Metadata metadata)
+        {
+            _ = metadata ?? throw new InvalidOperationException();
```
Nit - throw exceptions for public APIs/inputs and `Debug.Assert` for internal methods.

tg-msft commented on sdk/storage/Azure.Storage.Blobs.Cryptography/src/ClientSideDecryptionPolicy.cs:
```diff
@@ -0,0 +1,412 @@
+
+            using (var stream = new MemoryStream(Encoding.UTF8.GetBytes(encryptedDataString)))
+            {
+                var serializer = new DataContractJsonSerializer(typeof(EncryptionData));
```
We don't use `DataContractJsonSerializer` any more -- we use the `System.Text.Json` library now.  I don't think you need to fix this right now, but please file a bug on removing this.

tg-msft commented on sdk/storage/Azure.Storage.Blobs.Cryptography/src/ClientSideDecryptionPolicy.cs:
```diff
@@ -0,0 +1,412 @@
+                encryptionData = (EncryptionData)serializer.ReadObject(stream);
+            }
+
+            _ = encryptionData.ContentEncryptionIV ?? throw new NullReferenceException();
```
Don't throw `NullRefEx` yourself.  If this needs to be a runtime error, `InvalidOperationException` is preferred.  Move this to the Exceptions class.

tg-msft commented on sdk/storage/Azure.Storage.Blobs.Cryptography/src/ClientSideDecryptionPolicy.cs:
```diff
@@ -0,0 +1,412 @@
+        /// </summary>
+        /// <param name="encryptionData">The encryption data.</param>
+        /// <returns>Encryption key as a byte array.</returns>
+        private Memory<byte> GetContentEncryptionKey(EncryptionData encryptionData)
```
I'm not going to put the comment on every method, but anything that's got sync/async pairs and isn't public should be merged into a single implementation.  Happy to walk through that pattern for the first few on Teams together if that's helpful.

tg-msft commented on sdk/storage/Azure.Storage.Blobs.Cryptography/src/ClientSideDecryptionPolicy.cs:
```diff
@@ -0,0 +1,412 @@
+            else
+            {
+                key = this.KeyResolver?.Resolve(encryptionData.WrappedContentKey.KeyId)
+                    ?? throw EncryptionErrors.KeyNotFound(encryptionData.WrappedContentKey.KeyId);
```
Nit - the error kind of gets lost in here.  Would it be more obvious to add a separate `if (key == null)` check below?

tg-msft commented on sdk/storage/Azure.Storage.Blobs.Cryptography/src/ClientSideDecryptionPolicy.cs:
```diff
@@ -0,0 +1,412 @@
+                            aesProvider.Padding = PaddingMode.None;
+                        }
+
+                        return new RollingBufferStream(new CryptoStream(contentStream, aesProvider.CreateDecryptor(), CryptoStreamMode.Read), 10 * Constants.MB);
```
Style nit - long line, maybe pull some of these out into separate variables

tg-msft replied:
Add an entry in EncryptionConstants for the 10MB

tg-msft commented on sdk/storage/Azure.Storage.Blobs.Cryptography/src/ClientSideDecryptionPolicy.cs:
```diff
@@ -0,0 +1,412 @@
+
+        private static HttpRange ParseHttpRange(string serializedRange)
+        {
+            var rangeValues = serializedRange.Split('=')[1].Split(new char[] { '-' }, StringSplitOptions.RemoveEmptyEntries);
```
We should add a few more checks to avoid possible NullRefs.  And maybe move this into the HttpRange class.

jaschrep-msft replied:
#8007

tg-msft replied:
I don't think that change will happen this release though because we're locked down on API changes in Azure.Core.  A few extra size checks here would still be good in the meantime.

tg-msft commented on sdk/storage/Azure.Storage.Blobs.Cryptography/src/EncryptedBlobClient.cs:
```diff
@@ -0,0 +1,323 @@
+    /// Azure Storage block blobs with client-side encryption. See
+    /// <see cref="BlobClient"/> for more details.
+    /// </summary>
+    public class EncryptedBlobClient : BlobClient
```
I thought we were going to derive this from `BlockBlobClient` to leave future wiggle room?

jaschrep-msft replied:
Discussed offline. Leaving as is.

tg-msft commented on sdk/storage/Azure.Storage.Blobs.Cryptography/src/EncryptedBlobClient.cs:
```diff
@@ -0,0 +1,323 @@
+            IKeyEncryptionKey keyEncryptionKey = default,
+            IKeyEncryptionKeyResolver keyResolver = default,
+            string encryptionKeyWrapAlgorithm = default,
+            BlobClientOptions options = default)
```
We should move all of these extra properties onto an `EncryptedBlobClientOptions` type.  It'll simplify all the .ctor overloads and provide a consistent API with CPK.

jaschrep-msft replied:
I have finalized on separating `BlobClientOptions` and `ClientsideEncryptionOptions` due to technical concerns with storing "source options" on a client to be able to recreate their pipeline down the road.

tg-msft commented on sdk/storage/Azure.Storage.Blobs.Cryptography/src/EncryptedBlobClient.cs:
```diff
@@ -0,0 +1,323 @@
+        /// </param>
+        public EncryptedBlobClient(
+            Uri blobUri,
+            StorageSharedKeyCredential credential,
```
You have overloads for SharedKey and Token, but no overload with just Uri for SAS.

tg-msft commented on sdk/storage/Azure.Storage.Blobs.Cryptography/src/EncryptedBlobClient.cs:
```diff
@@ -0,0 +1,323 @@
+        internal EncryptedBlobClient(Uri blobUri, HttpPipeline pipeline)
+            : base(blobUri, pipeline)
+        {
+            //TODO make new pipeline for this one
```
?

jaschrep-msft replied:
Forgot to clean these comments up. This is waiting on another core feature, same as the extension method at the bottom of this file. Will comment out and leave a similar todo.

tg-msft commented on sdk/storage/Azure.Storage.Blobs.Cryptography/src/EncryptedBlobClient.cs:
```diff
@@ -0,0 +1,323 @@
+        /// <returns>Transformed content stream.</returns>
+        internal protected override (Stream, Metadata) TransformContent(Stream content, Metadata metadata)
+        {
+            metadata ??= new Dictionary<string, string>();
```
Should this be case insensitive?

jaschrep-msft replied:
Unnecessary but doesn't hurt and reinforces metadata behavior. Will add.

tg-msft commented on sdk/storage/Azure.Storage.Blobs.Cryptography/src/EncryptedBlobClient.cs:
```diff
@@ -0,0 +1,323 @@
+        #endregion ctors
+
+        /// <summary>
+        /// This method performs a transform on the content stream for uploads. It is a no-op by default.
```
This doesn't look like a no-op.

jaschrep-msft replied:
Copy/paste from the base method. Will fix.

tg-msft commented on sdk/storage/Azure.Storage.Blobs.Cryptography/src/EncryptedBlobClient.cs:
```diff
@@ -0,0 +1,323 @@
+                var serializer = new DataContractJsonSerializer(typeof(EncryptionData));
+                serializer.WriteObject(stream, encryptionData);
+                stream.Position = 0;
+                metadata.Add(EncryptionConstants.ENCRYPTION_DATA_KEY, new StreamReader(stream).ReadToEnd());
```
What if the metadata already has this key?  Do we want this to throw an Argument exception?

tg-msft replied:
Nit - use Stream.Seek instead of setting the Position to 0.

tg-msft replied:
Should serialization/deserialization methods move onto EncryptionData instead of being included in other methods?

jaschrep-msft replied:
Moved serialization over to system.text.json, as a bug was hit where it was easier to convert than to handle. Have made various organizational changes alongside that.

tg-msft commented on sdk/storage/Azure.Storage.Blobs.Cryptography/src/EncryptedBlobClient.cs:
```diff
@@ -0,0 +1,323 @@
+        /// <param name="content">Content to transform.</param>
+        /// <param name="metadata">Content metadata to transform.</param>
+        /// <returns>Transformed content stream.</returns>
+        internal protected override (Stream, Metadata) TransformContent(Stream content, Metadata metadata)
```
Don't we need an async version of this?  You're making a bunch of sync calls that have async equivalents that we should be using for async code paths.

jaschrep-msft replied:
I was going to say this method actually just does stream wrapping and doesn't actually make the async-friendly calls, but it turns out we do the key wrapping here and I don't think it can be moved. Will convert.

jaschrep-msft replied:
Wait. We'd need a rerelase of Azure.Storage.Blobs. This is an override from that package.

tg-msft replied:
We could make a blocking API call for now if we want to release this ahead of the next Blobs release.

tg-msft commented on sdk/storage/Azure.Storage.Blobs.Cryptography/src/EncryptedBlobClient.cs:
```diff
@@ -0,0 +1,323 @@
+            using (AesCryptoServiceProvider aesProvider = new AesCryptoServiceProvider() { Key = generatedKey })
+            {
+
+                var encryptionData = new EncryptionData()
```
This is 90% constants.  Should it be set automatically by the `EncryptionData` ctor?

jaschrep-msft replied:
Will make a special constructor. But deserialization tools will still require all fields to be settable.

jaschrep-msft replied:
Edit. Will create a helper method. There is code in here that never got its async branching that will need it.

tg-msft commented on sdk/storage/Azure.Storage.Blobs.Cryptography/src/EncryptedBlobClient.cs:
```diff
@@ -0,0 +1,323 @@
+                    },
+                    WrappedContentKey = new WrappedKey()
+                    {
+                        Algorithm = null, // TODO
```
Do we need to address this now?  If not, let's at least file a bug.

jaschrep-msft replied:
Yes. I completely missed this because my mock key wrapping service didn't care for it. This has been caught in keyvault integration tests.

tg-msft commented on sdk/storage/Azure.Storage.Blobs.Cryptography/src/EncryptedBlobClient.cs:
```diff
@@ -0,0 +1,323 @@
+                };
+
+                return (
+                    new RollingBufferStream(
```
Can we pull this out into a separate variable?  The implicit tuple syntax gets a little lost like this.

tg-msft commented on sdk/storage/Azure.Storage.Blobs.Cryptography/src/EncryptedBlobClient.cs:
```diff
@@ -0,0 +1,323 @@
+            return (encryptionStream, metadata);
+        }
+
+        private (Stream, EncryptionData) EncryptStream(Stream plaintext)
```
Nit - Would we want a 16th `Stream` class called `StorageEncryptionStream` that carries this data with it?

jaschrep-msft replied:
I'm a bit unenthused about that. It makes sense in the case of encryption, but not necessarily in the case of the generic TransformContent method that we override to inject this behavior.

tg-msft commented on sdk/storage/Azure.Storage.Blobs.Cryptography/src/EncryptedBlobClient.cs:
```diff
@@ -0,0 +1,323 @@
+                options = new BlobClientOptions();
+            }
+            options.AddPolicy(policy, position);
+            return options;
```
We shouldn't modify the user's ClientOptions this way.  We should pass this to https://github.com/Azure/azure-sdk-for-net/blob/master/sdk/storage/Azure.Storage.Common/src/StorageClientOptions.cs#L87 (augmenting it to take another policy).  You'd then need to call the base .ctor that takes a URI and a pipeline.

tg-msft commented on sdk/storage/Azure.Storage.Blobs.Cryptography/src/EncryptionConstants.cs:
```diff
@@ -0,0 +1,30 @@
+
+namespace Azure.Storage.Blobs.Specialized.Cryptography
+{
+    internal class EncryptionConstants
```
We use regular PascalCase for constants in .NET instead of this C style.

tg-msft replied:
Make this a static class

tg-msft commented on sdk/storage/Azure.Storage.Blobs.Cryptography/src/EncryptionErrors.cs:
```diff
@@ -0,0 +1,23 @@
+﻿using System;
```
Missing copyright header

tg-msft commented on sdk/storage/Azure.Storage.Blobs.Cryptography/src/EncryptionErrors.cs:
```diff
@@ -0,0 +1,23 @@
+
+namespace Azure.Storage.Blobs.Specialized.Cryptography
+{
+    internal class EncryptionErrors
```
Make this a static class

tg-msft commented on sdk/storage/Azure.Storage.Blobs.Cryptography/src/EncryptionErrors.cs:
```diff
@@ -0,0 +1,23 @@
+                "library does not support the given encryption algorithm.");
+
+        public static ArgumentException NoKeyAccessor()
+            => new ArgumentException("Key and KeyResolver cannot both be null.");
```
Nit - take these property names as parameters.  Pass them into the method using `nameof(...)` and then they'll be updated automatically when refactoring.

tg-msft replied:
Include the parameter names of everything if you're throwing ArgumentException

tg-msft commented on sdk/storage/Azure.Storage.Blobs.Cryptography/src/LengthLimitingStream.cs:
```diff
@@ -0,0 +1,87 @@
+        {
+            if (this._maxLength.HasValue)
+            {
+                return (int)Math.Min(count, this._maxLength.Value - this._bytesRead);
```
Nit - any reason not to use `Math.Clamp` for this part of it?

jaschrep-msft replied:
Math.Clamp doesn't add value I don't think. What would I set as a minimum value, zero? I'd rather throw if someone passes in a negative count, instead of successfully reading zero bytes. We are already set up to do that.

I guess we could rename this function from clamp, but it conceptually feels like a clamping function. Would `ImposeLimit(int count)` be better?

tg-msft commented on sdk/storage/Azure.Storage.Blobs.Cryptography/src/LengthLimitingStream.cs:
```diff
@@ -0,0 +1,87 @@
+
+        public override bool CanWrite => false;
+
+        public override long Length => this._maxLength.HasValue ? this._maxLength.Value : this._wrappedStream.Length;
```
Does `_wrappedStream.Length` work safely here?  Let's say I have a 20 byte stream, seek to position 10, then wrap it in a `LengthLimitingStream`.  Should the Length be 20 when I can only see 10 bytes of it?

jaschrep-msft replied:
Modified to always return _bytesRead.

tg-msft commented on sdk/storage/Azure.Storage.Blobs.Cryptography/src/LengthLimitingStream.cs:
```diff
@@ -0,0 +1,87 @@
+    /// <summary>
+    /// A read-only stream. Wraps a <see cref="Stream"/> with an upper limit on how many bytes can be read.
+    /// </summary>
+    internal class LengthLimitingStream : Stream
```
Nit - I don't love the name of this because the length limiting is completely optional.  It's more that this is a fixed readonly window of a Stream.

jaschrep-msft replied:
Do you have an alternative? This name makes sense to me; it applies an optional limit to the length of content this stream can return.

tg-msft commented on sdk/storage/Azure.Storage.Blobs.Cryptography/src/Models/EncryptionAlgorithm.cs:
```diff
@@ -0,0 +1,18 @@
+    /// <summary>
+    /// Specifies the encryption algorithm used to encrypt a resource.
+    /// </summary>
+    public enum ClientsideEncryptionAlgorithm
```
Why isn't this just `EncryptionAlgorithm`?

jaschrep-msft replied:
EncryptionAlgorithm is an overloaded term when CPK also exists in the main library. EncryptionAlgorithm is an already existing enum defined by swagger and I wanted to ensure the two aren't confused.

tg-msft replied:
It'd be nice of all the encryption names matched up in intellisense so let's try to get this aligned to whatever we call `ClientsideEncryptionOptions`.

tg-msft replied:
Also, should this be an extensible enum?

tg-msft commented on sdk/storage/Azure.Storage.Blobs.Cryptography/src/Models/EncryptionAlgorithm.cs:
```diff
@@ -0,0 +1,18 @@
+        /// <summary>
+        /// AES-CBC using a 256 bit key.
+        /// </summary>
+        AES_CBC_256
```
Can we make this look more like the naming in KeyVault?   https://github.com/Azure/azure-sdk-for-net/blob/master/sdk/keyvault/Microsoft.Azure.KeyVault.Cryptography/src/Algorithms/Aes128CbcHmacSha256.cs#L10

jaschrep-msft replied:
I'm not sure which part(s) of that you're going for.

tg-msft replied:
Using the name `AesCbc256`.

tg-msft commented on sdk/storage/Azure.Storage.Blobs.Cryptography/src/Models/EncryptionAgent.cs:
```diff
@@ -0,0 +1,26 @@
+    /// Represents the encryption agent stored on the service.
+    /// </summary>
+    [DataContract]
+    public struct EncryptionAgent
```
Should this be public?  Isn't there only one accepted value for this?

jaschrep-msft replied:
It's being serialized. The serialization package needs to be able to see the class.

tg-msft commented on sdk/storage/Azure.Storage.Blobs.Cryptography/src/Models/EncryptionData.cs:
```diff
@@ -0,0 +1,60 @@
+    /// Represents the encryption data that is stored on the service.
+    /// </summary>
+    [DataContract]
+    public struct EncryptionData
```
Mutable structs that contain other mutable structs are a bad time.  This should be a class.

tg-msft replied:
Does this need to be public in v1?

jaschrep-msft replied:
It's being serialized. The serialization package needs to be able to see the class.

tg-msft commented on sdk/storage/Azure.Storage.Blobs.Cryptography/src/Models/EncryptionData.cs:
```diff
@@ -0,0 +1,60 @@
+        public string EncryptionMode { get; set; }
+
+        /// <summary>
+        /// A {@link WrappedKey} object that stores the wrapping algorithm, key identifier and the encrypted key.
```
Nit - java doc comment

tg-msft commented on sdk/storage/Azure.Storage.Blobs.Cryptography/src/Models/EncryptionData.cs:
```diff
@@ -0,0 +1,60 @@
+        /// The blob encryption mode.
+        /// </summary>
+        [DataMember]
+        public string EncryptionMode { get; set; }
```
If the type has to be public, do the setters at least?

tg-msft replied:
Should this be an enum?

jaschrep-msft replied:
My understanding is that serialization packages need access to setters to deserialize into these objects. Unless I write my own custom deserializer for the whole thing, in which case why am I even using the deserialization package?

tg-msft commented on sdk/storage/Azure.Storage.Blobs.Cryptography/src/Models/EncryptionData.cs:
```diff
@@ -0,0 +1,60 @@
+        /// Metadata for encryption. Currently used only for storing the encryption library, but may contain other data.
+        /// </summary>
+        [DataMember]
+        public Metadata KeyWrappingMetadata { get; set; }
```
Should the naming pattern match `WrappedContentKey`?

jaschrep-msft replied:
This is the json key exactly.

tg-msft commented on sdk/core/Azure.Core/src/Http/HttpHeader.cs:
```diff
@@ -58,6 +58,7 @@ public static class Names
             public static string Accept => "Accept";
             public static string Authorization => "Authorization";
             public static string Range => "Range";
+            public static string XMsRange => "x-ms-range";
```
We should move this to a EncryptionConstants for now because it's causing the build to fail since Azure.Core just shipped.

tg-msft commented on sdk/storage/Azure.Storage.Blobs.Cryptography/src/Models/WrappedKey.cs:
```diff
@@ -0,0 +1,43 @@
+        /// Base64-encoded content encryption IV.
+        /// </summary>
+        [DataMember(Name = "EncryptedKey")]
+        public string Base64ContentEncryptionIV
```
Copy/paste error?

tg-msft commented on sdk/storage/Azure.Storage.Blobs/src/BlobBaseClient.cs:
```diff
@@ -1227,8 +1227,18 @@ Task<Response<BlobProperties>> GetPropertiesAsync(bool async, CancellationToken
-                Response<BlobDownloadInfo> response = await client.DownloadAsync(accessConditions: blobAccessConditions, cancellationToken: ct).ConfigureAwait(false);
-                await response.Value.Content.CopyToAsync(destination, 81920 /* default value */, ct).ConfigureAwait(false);
+                Response<BlobDownloadInfo> response = await client.DownloadInternal(
+                    default, blobAccessConditions, default, async, cancellationToken).ConfigureAwait(false);
```
Nit - please used a named parameter or add a `/* name */` comment if you're passing just `default` in a big argument list.

tg-msft commented on sdk/storage/Azure.Storage.Blobs/src/BlobBaseClient.cs:
```diff
@@ -1215,10 +1215,10 @@ private void SetNameFieldsIfNull()
-                        cancellationToken: ct
-                        );
+                        async: async,
+                        cancellationToken: ct);
```
Thanks for fixing this part of it.  Ideally we'd also make this an `async` function and add a `ConfigureAwait(false)`.  Errors flow better: https://github.com/davidfowl/AspNetCoreDiagnosticScenarios/blob/master/AsyncGuidance.md#prefer-asyncawait-over-directly-returning-task

tg-msft commented on sdk/storage/Azure.Storage.Blobs/src/RollingBufferStream.cs:
```diff
@@ -0,0 +1,213 @@
+using System;
+using System.IO;
+
+namespace Azure.Storage.Blobs
```
Does this have to be in `Blobs` or can it move into the `Blobs.Cryptography` assembly?

jaschrep-msft replied:
It doesn't have to be, but I feel like it's general purpose enough that it deserves to be accessible if/when other packages need it. Honestly, looking at it again, I feel it should be in `Common`.

tg-msft replied:
I'd rather move it when other packages actually need it.  This should move into the Crypto packages.

tg-msft commented on sdk/storage/Azure.Storage.Blobs/src/RollingBufferStream.cs:
```diff
@@ -0,0 +1,213 @@
+        ///// <summary>
+        ///// The index where the sequential data begins.
+        ///// </summary>
+        //private int _rollingBufferBreakPos = 0;
```
Remove if not needed?

tg-msft commented on sdk/storage/Azure.Storage.Blobs/src/RollingBufferStream.cs:
```diff
@@ -0,0 +1,213 @@
+            {
+                if (value < 0)
+                {
+                    throw new ArgumentException("Cannot seek to a negative position");
```
Move these to the Errors class

tg-msft commented on sdk/storage/Azure.Storage.Blobs/src/RollingBufferStream.cs:
```diff
@@ -0,0 +1,213 @@
+             */
+        }
+
+        public override int Read(byte[] buffer, int offset, int count)
```
We need an async implementation too or you'll introduce blocking into async code paths _(the base ReadAsync will call into Read if it hasn't been overridden)_.  Ideally we'd share it with a `bool async` helper.

tg-msft commented on sdk/storage/Azure.Storage.Common/src/Errors.cs:
```diff
@@ -70,6 +70,9 @@ public static ArgumentException AccountMismatch(string accountNameCredential, st
             => new ArgumentException("Connection string parsing error");
 
+        public static ArgumentException ParsingHttpRangeFailed()
+            => new ArgumentException("Could not parse the serialized range.");
```
Nit - include the argument name

tg-msft commented on sdk/storage/Azure.Storage.Blobs/tests/EncryptedBlockBlobClientTests.cs:
```diff
@@ -18,25 +25,281 @@ public EncryptedBlockBlobClientTests(bool async)
+        {
+            var disposable = this.GetNewContainer(out var container);
+            blob = new EncryptedBlobClient(
+                    new Uri(Path.Combine(container.Uri.ToString(), this.GetNewBlobName())), this.GetNewSharedKeyCredentials(),
```
Nit - use `BlobUriBuilder` or even `container.GetBlobClient(GetNewBlobName()).Uri`.

tg-msft replied:
Nit - put each parameter on its own line

tg-msft replied:
Ideally we'd create the `BlobServiceClient` using these credentials so `GetNewContainer` and `GetNewSharedKeyCredentials` never get out of alignment.

jaschrep-msft replied:
I don't understand this last comment.

tg-msft commented on sdk/storage/Azure.Storage.Blobs/tests/EncryptedBlockBlobClientTests.cs:
```diff
@@ -18,25 +25,281 @@ public EncryptedBlockBlobClientTests(bool async)
+        /// </summary>
+        /// <param name="metadata">The blob's metadata</param>
+        /// <returns>The relevant metadata.</returns>
+        private EncryptionData GetAndValidateEncryptionData(IDictionary<string, string> metadata)
```
Is this duplicated from the code?  Just make it internal and expose it via InternalsVisibleTo?

tg-msft commented on sdk/storage/Azure.Storage.Blobs/tests/EncryptedBlockBlobClientTests.cs:
```diff
@@ -18,25 +25,281 @@ public EncryptedBlockBlobClientTests(bool async)
+        [TestCase(Constants.KB)] // multiple blocks
+        [TestCase(Constants.KB - 4)] // multiple unalligned blocks
+        [LiveOnly]
+        public void Upload(long dataSize)
```
We only have to write Async tests.  The framework magically converts them to sync tests for us (if you call InstrumentClient, etc.).  Sean can give you a rundown.

tg-msft commented on sdk/storage/Azure.Storage.Blobs/tests/EncryptedBlockBlobClientTests.cs:
```diff
@@ -18,25 +25,281 @@ public EncryptedBlockBlobClientTests(bool async)
+        private IDisposable GetEncryptedBlockBlobClient(out EncryptedBlobClient blob, MockKeyEncryptionKey mock)
+        {
+            var disposable = this.GetNewContainer(out var container);
+            blob = new EncryptedBlobClient(
```
Nit - wrap this in `InstrumentClient`

tg-msft commented on sdk/storage/Azure.Storage.Blobs/tests/EncryptedBlockBlobClientTests.cs:
```diff
@@ -18,25 +25,281 @@ public EncryptedBlockBlobClientTests(bool async)
+            blob = new EncryptedBlobClient(
+                    new Uri(Path.Combine(container.Uri.ToString(), this.GetNewBlobName())), this.GetNewSharedKeyCredentials(),
+                    keyEncryptionKey: mock,
+                    keyResolver: mock);
```
You also need our standard test BlobClientOptions passed in: https://github.com/Azure/azure-sdk-for-net/blob/master/sdk/storage/Azure.Storage.Blobs/tests/BlobTestBase.cs#L46

tg-msft commented on sdk/storage/Azure.Storage.Blobs/tests/EncryptedBlockBlobClientTests.cs:
```diff
@@ -18,25 +25,281 @@ public EncryptedBlockBlobClientTests(bool async)
+        [TestCase(14)] // a single unalligned cipher block
+        [TestCase(Constants.KB)] // multiple blocks
+        [TestCase(Constants.KB - 4)] // multiple unalligned blocks
+        [LiveOnly]
```
We need recorded tests too.  Sean can help you get that setup.

jaschrep-msft replied:
The client functionality being tested securely generates keys using `System.Security.Cryptography.RNGCryptoServiceProvider`. I don't know if its possible to record that. Sean isn't either.

tg-msft commented on sdk/storage/Azure.Storage.Blobs/tests/MockKeyEncryptionKey.cs:
```diff
@@ -0,0 +1,87 @@
+    /// <summary>
+    /// Mock for a key encryption key. Not meant for production use.
+    /// </summary>
+    internal class MockKeyEncryptionKey : IKeyEncryptionKey, IKeyEncryptionKeyResolver
```
Have we tested this against KeyVault's Track 2 library?  We should really get that checked in somewhere so we're validating the only meaningful customer scenario end to end.

tg-msft replied:
Have we tested this against Track 1 for that matter?  Or other languages?  It'd be great to have confidence in the interop story before we ship.

jaschrep-msft replied:
We have not tested against keyvault yet as it was still under development when writing this.

We have not run backwards compatible tests yet, though that is a requirement before shipping. Unfortunately, since this touches files outside of the crypto package, we need to get those changes in before main package release, even if we delay the crypto package release to accommodate those tests.

tg-msft commented on sdk/storage/Azure.Storage.Blobs/tests/RollingBufferStreamTests.cs:
```diff
@@ -0,0 +1,147 @@
+    /// <summary>
+    /// Read-only, unseekable Stream where the 0-indexed nth byte read will have the value n.
+    /// </summary>
+    internal class MockStream : Stream
```
Pull this out into its own file.  We allow extension methods to break the "one type per file" rule, but not whole types like this.

tg-msft commented on sdk/storage/Azure.Storage.sln:
```diff
@@ -105,9 +105,9 @@ Project("{2150E333-8FDC-42A3-9474-1A3956D46DE8}") = "Cryptography", "Cryptograph
 EndProject
 Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "Azure.Storage.Files.DataLake", "Azure.Storage.Files.DataLake\src\Azure.Storage.Files.DataLake.csproj", "{DE6EC5B8-5DBA-4D21-931E-3B042010B3C0}"
 EndProject
-Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "Azure.Storage.Files.DataLake.Tests", "Azure.Storage.Files.DataLake\tests\Azure.Storage.Files.DataLake.Tests.csproj", "{EDBF4B6B-5D28-47AF-A9BE-B70568067407}"
```
And you said you were going to remove `Queues.Cryptography` as part of this change?

tg-msft commented on sdk/storage/Azure.Storage.Blobs/src/BlobClient.cs:
```diff
@@ -1057,5 +1063,16 @@ async Task<Response<BlobContentInfo>> UploadStreamAsync()
+        /// <param name="content">Content to transform.</param>
+        /// <param name="metadata">Content metadata to transform.</param>
+        /// <returns>Transformed content stream and metadata.</returns>
+        internal protected virtual (Stream, Metadata) TransformContent(Stream content, Metadata metadata)
```
Remove the `protected` - I'm not sure we want this to be a public API without more consideration.

jaschrep-msft replied:
Oh! I had always thought `internal protected` was the intersection of that visibility, not the union. Will fix.

tg-msft commented on sdk/storage/Azure.Storage.Blobs/src/BlobClient.cs:
```diff
@@ -824,8 +824,10 @@ bool TryGetStreamLength(Stream stream, out long length)
-                =>
-                client.UploadInternal(
+            {
+                (content, metadata) = TransformContent(content, metadata);
```
How does this work with retry policies?  Will we potentially transform the content more than once?

jaschrep-msft replied:
I have not tested with forced retries. This is still within the convenience layer, though, so we can't end up calling this method more than once.

The implementation we use is not seekable, so we will throw before we attempt to transform twice. Will run tests and follow up on this.

tg-msft commented on sdk/storage/Azure.Storage.Blobs.Cryptography/src/Models/EncryptedBlobRange.cs:
```diff
@@ -0,0 +1,73 @@
+    /// <summary>
+    /// This is a representation of a range of bytes on an encrypted blob, which may be expanded from the requested range to
+    /// included extra data needed for encryption.Note that this type is not strictly thread-safe as the download method
+    /// will update the count in case the user did not specify one. Passing null as an EncryptedBlobRange value will default
```
> the download method will update the count in case the user did not specify one

Where is this happening?  I think I lost it in the PR.  Does it muck directly with the value?  Could we add an `UpdateRangeLength` method or something to make it a little more clear?

jaschrep-msft replied:
I think the docstring is no longer accurate. I copy/pasted from Java's code before modifying the code. Will clarify.

tg-msft reviewed (commented):
I added a bunch of comments, but clearly everything can't be fixed today.  My major concerns are:

- The namespace being `.Specialized.Cryptography` vs. `.Specialized`
- Changing the base class from `BlockBlobClient` to `BlobClient`
- Adding a bunch of .ctor parameters instead of an `EncryptedBlobClientOptions` and mucking with the user's policies
- The public API addition to `BlobClient` that I think we want to hide for now
- Verification that we've tested this with KeyVault's Track 2 library

jaschrep-msft commented on sdk/storage/Azure.Storage.Blobs.Cryptography/src/ClientSideDecryptionPolicy.cs:
```diff
@@ -0,0 +1,393 @@
+    /// pipes the returned content stream through a decryption stream. Should only be in pipelines for instances
+    /// of <see cref="EncryptedBlobClient"/>
+    /// </summary>
+    public class ClientSideDecryptionPolicy : HttpPipelinePolicy
```
I've been told all policies need to be public to support people building their own pipelines using our tools. Is this not the case?

tg-msft replied:
We're making more policies internal and letting them be controlled via ClientOptions to limit API scope _(i.e., it's always easier to make things public when customers have an explicit need than change an API)_.

jaschrep-msft commented on sdk/storage/Azure.Storage.Blobs.Cryptography/src/ClientSideDecryptionPolicy.cs:
```diff
@@ -0,0 +1,393 @@
+using Azure.Core.Pipeline;
+using Metadata = System.Collections.Generic.IDictionary<string, string>;
+
+namespace Azure.Storage.Blobs.Specialized.Cryptography
```
I understood we wanted to be under `Specialized` and also a `Cryptography` namespace. This is how I interpreted that.

While I agree it might be a little redundant to have one extra layer to the namespace, given the user already opting into the code by downloading the package in the first place, I also feel it's unlikely the code using this cryptography client will want anything to do with block, page, and append blob clients. Users probably don't want to clog intellisense with that when they just want encryption and don't care about the technical structure of azure storage. This extra layer gives them that option.

tg-msft replied:
Anyone using this feature is advanced enough that they can handle a few other clients in the namespace.  Also, the Arch Board already signed off on `.Specialized` and we'll need to go back to them (per https://azure.github.io/azure-sdk/dotnet_introduction.html#dotnet-namespace-naming) if we want to change it.  I'd prefer `.Specialized`.

jaschrep-msft commented on sdk/storage/Azure.Storage.Blobs.Cryptography/src/ClientSideDecryptionPolicy.cs:
```diff
@@ -0,0 +1,393 @@
+        private IKeyEncryptionKey LocalKey { get; }
+
+        /// <summary>
+        /// Initializes a new instance of the {@link BlobEncryptionPolicy} class with the specified key and resolver.
```
Thanks for catching the copy/paste mistake. Will check for all of these.

jaschrep-msft commented on sdk/storage/Azure.Storage.Blobs.Cryptography/src/ClientSideDecryptionPolicy.cs:
```diff
@@ -0,0 +1,393 @@
+            if (message.Request.Headers.TryGetValue(HttpHeader.Names.Range, out var range))
+            {
+                encryptedRange = new EncryptedBlobRange(ParseHttpRange(range));
+                message.Request.Headers.Add(encryptedRange.AdjustedRange.ToRangeHeader());
```
I want SetValue.

jaschrep-msft commented on sdk/storage/Azure.Storage.Blobs.Cryptography/src/ClientSideDecryptionPolicy.cs:
```diff
@@ -0,0 +1,393 @@
+
+            // parse header value (e.g. "bytes <start>-<end>/<blobSize>")
+            // end is the inclusive last byte; e.g. header "bytes 0-7/8" is the entire 8-byte blob
+            var tokens = contentRange.Split(new char[] { ' ', '-', '/' }); // ["bytes", "<start>", "<end>", "<blobSize>"]
```
HttpRange only has a ToString, no Parse/TryParse. It definitely should, though.

tg-msft replied:
Filed #8007 to add this in Azure.Core

jaschrep-msft committed:
Addressed Comments

jaschrep-msft committed:
Moved constant

jaschrep-msft committed:
Additional Tests

Tests track 1 compatibility.
Tests KeyVault compatibility.

jaschrep-msft committed:
Fixed Track 1 compatibility bugs

jaschrep-msft commented:
Now fixes #8001

tg-msft commented on sdk/storage/Azure.Storage.Blobs.Cryptography/src/Models/EncryptionData.cs:
```diff
@@ -1,59 +1,56 @@
+        /// <param name="json">JSON to deserialize.</param>
+        /// <returns></returns>
+        public static EncryptionData Deserialize(string json)
+            => JsonSerializer.Deserialize<EncryptionData>(json);
```
We're not using the higher level `Deserialize`/`Serialize` types for performance and compat.  If you use this, this library won't run in an Azure Function (until they update the version of some of their components).

jaschrep-msft replied:
Can you point me to a resource that demonstrates the methods you like?

jaschrep-msft committed:
Addressed Comments

jaschrep-msft committed:
Reverted test hack

jaschrep-msft committed:
Merge branch 'master' of https://github.com/Azure/azure-sdk-for-net into storage/feature/clientside-encryption

HengzheLi commented:
Can I assume that the latest SDK version (>=12.0) only support client-side encryption as before versions' (<=11) after this PR is merged.

jaschrep-msft commented:
@HengzheLi this is correct. While versions >= 12.0.0 support [server-side encryption with customer provided keys](https://azure.microsoft.com/en-us/blog/customer-provided-keys-with-azure-storage-service-encryption/), we do not yet have client-side encryption support in this major version.

jaschrep-msft committed:
Merge branch 'master' into storage/feature/clientside-encryption

jaschrep-msft committed:
Fix silent merge issues in sln/csproj files

jaschrep-msft committed:
Added Azure.Core funcitonality for cryptography

Azure core now supports Azure.Storage.Blobs.Cryptography.
Still failing api compat.

jaschrep-msft committed:
Clientside encryption functioning

API compat temporarily disabled due to false positive.
Reinserted TransformContent into BlobClient.
Fixed issue with async branching in blob decryption.

jaschrep-msft committed:
Client creation modified for clientside encryption

Constructors now all pass around their source options and auth policies
Bug in clientside encryption tests fixed

jaschrep-msft commented:
A few conceptual changes to how clients are structured in Azure.Storage.Blobs:
1. All clients now save an internally accessible `BlobClientOptions` instance, a deep copy of the options used to make the client's pipeline. This supports the "clone with modification" pipeline behavior necessary for this feature.
2. All clients now save an internally accessible `HttpPipelinePolicy` instance, a reference to the authentication policy living in the client's pipeline. Access to this policy is necessary to build a pipeline from the saved options.

Notable changes/behavior:
1. Once the decryption policy gets into the EncryptedBlobClient pipeline, there is no code to get it out. However, I am not aware of a method for a BaseBlobClient to go back to container or service clients, where they would become an issue.
2. Internal constructor signatures have been significantly altered. Each client now only has a single constructor implementation (instead of 2-3), with the rest of the constructors forwarding through other signatures. Comments have been left explaining when non-obvious argument transforms take place.

jaschrep-msft commented:
Additionally, API compat was disabled in this branch due to a false positive.

jaschrep-msft committed:
Merge branch 'master' of https://github.com/Azure/azure-sdk-for-net into storage/feature/clientside-encryption

Api compat error (even though it's still turned off in core .csproj

jaschrep-msft committed:
Export API

jaschrep-msft commented:
/azp run net - storage -ci

jaschrep-msft commented on sdk/storage/Azure.Storage.Blobs.Cryptography/src/ClientSideDecryptionPolicy.cs:
```diff
@@ -0,0 +1,414 @@
+        public override void Process(HttpMessage message, ReadOnlyMemory<HttpPipelinePolicy> pipeline)
+        {
+            ProcessImpl(message, pipeline, false).EnsureCompleted();
+            //EncryptedBlobRange encryptedRange = default;
```
Will remove

jaschrep-msft committed:
Addressed comments

jaschrep-msft commented:
Splitting the exposed `TransformContent` into sync and async overloads seems to have upset the test interceptor only for the `UploadStreamAsync_SmallBlobs` test.

jaschrep-msft commented:
Calling out a new public API in storage code:
```csharp
protected virtual (Stream, Metadata) TransformContent(Stream content, Metadata metadata);
protected virtual Task<(Stream, Metadata)> TransformContentAsync(Stream content, Metadata metadata)
```
This method is a no-op by default, and is exposed so that the encryption package can override it with its cryptographic transformation.

jaschrep-msft committed:
Reworked TransformContent signature

jaschrep-msft committed:
Export-API

jaschrep-msft commented:
/azp run net - storage -ci

jaschrep-msft committed:
Test infra patches

jaschrep-msft commented:
/azp run net - storage -ci

jaschrep-msft committed:
Moved UploadContent to its own file

tg-msft commented on sdk/storage/Azure.Storage.Blobs.Cryptography/src/ClientsideEncryptionOptions.cs:
```diff
@@ -0,0 +1,35 @@
+    /// <summary>
+    /// Provides the client configuration options for connecting to Azure Blob using clientside encryption.
+    /// </summary>
+    public class ClientsideEncryptionOptions
```
`Clientside` reads a little weird to me (should that be two words?).  What did they call this in Track 1/Java Track 2?  Maybe something morel like `EncryptedBlobOptions` to match the `EncryptedBlobClient` name?

jaschrep-msft replied:
It could be two words. Using the track 1 name is not necessarily the best metric, as CPK didn't exist in the track 1 world. A few names have been selected to be distinct from CPK naming, as some model names for that feature use a generic "Encryption_____" naming pattern. I'll start a thread offline for this naming decision.

tg-msft commented on sdk/storage/Azure.Storage.Blobs.Cryptography/src/EncryptionErrors.cs:
```diff
@@ -0,0 +1,27 @@
+
+        public static ArgumentException BadEncryptionAlgorithm()
+            => new ArgumentException("Invalid Encryption Algorithm found on the resource. This version of the client" +
+                "library does not support the given encryption algorithm.");
```
Can we give the name of the algorithm we don't understand as a bit of help to the user?

tg-msft commented on sdk/storage/Azure.Storage.Blobs.Cryptography/src/EncryptionErrors.cs:
```diff
@@ -0,0 +1,27 @@
+                "library does not support the given encryption algorithm.");
+
+        public static ArgumentException NoKeyAccessor(params string[] names)
+            => new ArgumentException($"One of [{string.Join(",", names)}] must be non-null.");
```
Nit - `", "` will read a little nicer than `","`

tg-msft commented on sdk/storage/Azure.Storage.Blobs.Cryptography/src/Models/EncryptionAgent.cs:
```diff
@@ -0,0 +1,21 @@
+        /// <summary>
+        /// The algorithm used for encryption.
+        /// </summary>
+        public string EncryptionAlgorithm { get; set; }
```
Should this be a `ClientsideEncryptionAlgorithm` type?

tg-msft commented on sdk/storage/Azure.Storage.Blobs/tests/Azure.Storage.Blobs.Tests.csproj:
```diff
@@ -8,9 +8,10 @@
     <ProjectReference Include="$(MSBuildThisFileDirectory)..\src\Azure.Storage.Blobs.csproj" />
     <ProjectReference Include="$(MSBuildThisFileDirectory)..\..\Azure.Storage.Common\src\Azure.Storage.Common.csproj" />
-    <ProjectReference Include="$(MSBuildThisFileDirectory)..\..\Azure.Storage.Blobs.Cryptography\src\Azure.Storage.Blobs.Cryptography.csproj" />
+    <ProjectReference Include="$(MSBuildThisFileDirectory)..\..\..\identity\Azure.Identity\src\Azure.Identity.csproj" />
```
Do we need a project reference to Identity when we have a package reference above?

jaschrep-msft replied:
No. Probably just the result of some mistakes with the VS UI.

tg-msft commented on sdk/storage/Azure.Storage.Blobs/tests/BlobClientTests.cs:
```diff
@@ -573,20 +527,6 @@ private static async Task DownloadAndAssertAsync(Stream stream, BlobClient blob)
             // Use a 1KB threshold so we get a lot of individual blocks
             await UploadStreamAndVerify(size, Constants.KB, new StorageTransferOptions { MaximumTransferLength = Constants.KB });
 
-        [Test]
```
Why are you removing all these tests?   We should probably revert all the changes in this file.

jaschrep-msft replied:
I identified an unused internal method in the SDK, and these tests are for that method. I discussed removing that method with you and you were onboard.

jaschrep-msft commented on eng/Directory.Build.Data.targets:
```diff
@@ -88,7 +88,7 @@
     <CentralPackagesFile>$(MSBuildThisFileDirectory)Packages.Data.props</CentralPackagesFile>
     <CentralPackageVersionPackagePath>$(MSBuildThisFileDirectory)Microsoft.Build.CentralPackageVersions\2.0.46\Sdk</CentralPackageVersionPackagePath>
-    <UseProjectReferenceToAzureClients Condition="'$(UseProjectReferenceToAzureClients)' == ''">false</UseProjectReferenceToAzureClients>
+    <UseProjectReferenceToAzureClients Condition="'$(UseProjectReferenceToAzureClients)' == ''">true</UseProjectReferenceToAzureClients>
```
This needed to be flipped for CI when the azure core changes were still on the menu, and I was not sure of how to unflip it until the core change was published to nuget. This should be reverted along with the other azure core changes. Will handle.

tg-msft commented on sdk/storage/Azure.Storage.Blobs/src/Models/UploadContent.cs:
```diff
@@ -0,0 +1,24 @@
+    /// <summary>
+    /// Blob content to be uploaded by a <see cref="Specialized.BlobBaseClient"/>.
+    /// </summary>
+    public class UploadContent
```
We should probably name this `BlobUploadContent` because we've been burned a couple of times when we haven't put the `Blob` prefix in front of things that we eventually want to add to Queues/Files/DataLake down the road.

tg-msft commented on sdk/storage/Azure.Storage.Blobs/src/Models/UploadContent.cs:
```diff
@@ -0,0 +1,24 @@
+        /// <summary>
+        /// Content stream to upload.
+        /// </summary>
+        public Stream ContentStream { get; set; }
```
We should call this just `Content` to be consistent with other `Stream content` parameters, `BlobDownloadInfo.Content`, etc.

jaschrep-msft commented on sdk/core/Azure.Core/src/Azure.Core.csproj:
```diff
@@ -9,6 +9,7 @@
     <TargetFrameworks>$(RequiredTargetFrameworks)</TargetFrameworks>
     <AllowUnsafeBlocks>true</AllowUnsafeBlocks>
     <EnableClientSdkAnalyzers>false</EnableClientSdkAnalyzers>
+    <EnableApiCompat>false</EnableApiCompat>
```
@pakrym recommended this as a workaround for what he confirmed as a false positive, and this was necessary to pass CI. The false positive was to do with the changes in core we decided to revert, though, so we can turn it on with that revert, instead of waiting until everything else was approved and passing.

jaschrep-msft commented on sdk/storage/Azure.Storage.Blobs.Cryptography/src/BlobClientOptionsExtensions.cs:
```diff
@@ -0,0 +1,23 @@
+        {
+            options = options == default
+                 ? new BlobClientOptions()
+                 : new BlobClientOptions(options);
```
All the options saving work needs to be reverted since Core won't ship the changes making it worth it.

tg-msft commented on sdk/storage/Azure.Storage.Blobs/src/BlobServiceClient.cs:
```diff
@@ -41,6 +41,19 @@ public class BlobServiceClient
+        /// </summary>
+        internal virtual BlobClientOptions SourceOptions => _sourceOptions == default
+            ? default
+            : new BlobClientOptions(_sourceOptions);
```
We shouldn't create new instances every time the property is accessed.

jaschrep-msft replied:
We should. This is a mutable class. Allowing ourselves to edit this breaks assumptions about what the field represents.

However, without the Azure.Core changes, we can't create a new instance anymore, so we're going to have to live with mutating these values for now.

tg-msft commented on sdk/storage/Azure.Storage.Blobs.Cryptography/src/ClientSideDecryptionPolicy.cs:
```diff
@@ -0,0 +1,373 @@
+    /// pipes the returned content stream through a decryption stream. Should only be in pipelines for instances
+    /// of <see cref="EncryptedBlobClient"/>
+    /// </summary>
+    internal class ClientSideDecryptionPolicy : HttpPipelinePolicy
```
Nit - the casing of `ClientSide` doesn't match the other `Clientside`s, but I think we should match whatever pattern we follow for all of those names.

tg-msft commented on sdk/storage/Azure.Storage.Blobs.Cryptography/src/ClientSideDecryptionPolicy.cs:
```diff
@@ -0,0 +1,373 @@
+            await ProcessImpl(message, pipeline, false).ConfigureAwait(false);
+        }
+
+        private async ValueTask ProcessImpl(HttpMessage message, ReadOnlyMemory<HttpPipelinePolicy> pipeline, bool async)
```
Slight nit - this is kind of dumb, but we have an unofficial guideline to not use `Impl` as a suffix because it makes stack traces ugly.  Would you mind switching to `Internal` as a suffix instead?

tg-msft commented on sdk/storage/Azure.Storage.Blobs.Cryptography/src/ClientSideDecryptionPolicy.cs:
```diff
@@ -0,0 +1,373 @@
+            EncryptedBlobRange encryptedRange = default;
+            if (message.Request.Headers.TryGetValue(HttpHeader.Names.Range, out var range))
+            {
+                encryptedRange = new EncryptedBlobRange(ParseHttpRange(range));
```
Is this idempotent (i.e., if this ends up being retried)?

jaschrep-msft replied:
This is a per-call policy. It shouldn't be retried.

jaschrep-msft replied:
But no. We will keep making the start of the range earlier and earlier.

tg-msft commented on sdk/storage/Azure.Storage.Blobs.Cryptography/src/ClientSideDecryptionPolicy.cs:
```diff
@@ -0,0 +1,373 @@
+            // parse header value (e.g. "bytes <start>-<end>/<blobSize>")
+            // end is the inclusive last byte; e.g. header "bytes 0-7/8" is the entire 8-byte blob
+            var tokens = contentRange.Split(new char[] { ' ', '-', '/' }); // ["bytes", "<start>", "<end>", "<blobSize>"]
+            if (tokens.Length < 4)
```
We're expecting exactly 4, right?  Should we strengthen this check?

tg-msft commented on sdk/storage/Azure.Storage.Blobs.Cryptography/src/ClientSideDecryptionPolicy.cs:
```diff
@@ -0,0 +1,373 @@
+
+        private static Metadata ExtractMetadata(ResponseHeaders headers)
+        {
+            Metadata metadata = new Dictionary<string, string>(StringComparer.OrdinalIgnoreCase);
```
Do we need all the metadata?  Could we just look up the one or two keys we use?

jaschrep-msft committed:
Core changes reverted and custom serializer added

jaschrep-msft commented on sdk/storage/Azure.Storage.Common/tests/Shared/KeyVaultConfiguration.cs:
```diff
@@ -0,0 +1,40 @@
+    public class KeyVaultConfiguration
+    {
+        public string VaultName { get; private set; }
+        public string VaultEndpoint { get; private set; }
```
Vault name is not the account name. It's just the helpful name of the vault in the test config file so the top-level config can refer to it with that name. It in no way relates to the name of the keyvault itself, and Storage tenant configs have a similar structure.

jaschrep-msft commented on sdk/storage/Azure.Storage.Blobs.Cryptography/src/Models/WrappedKey.cs:
```diff
@@ -0,0 +1,26 @@
+        /// <summary>
+        /// The algorithm used for wrapping.
+        /// </summary>
+        public string Algorithm { get; set; }
```
This is a user-provided string that will be passed in as the algorithm name to the user-provided IKeyEncryptionKey::WrapKey(Async). We have no default or suggested values for this, so I don't think an enum buys anything. I could clarify the docstrings here and where the user passes the value in.

jaschrep-msft committed:
Addressed Comments

Updated value for encryption agent in encryption metadata.
Removed Azure.Identity and Azure.Security.Keys.Kevault from Azure.Storage.sln

Minor error message changes.
Minor type/member name changes.

jaschrep-msft commented:
With the vast internal design changes through this PR's history, alongside new work to support the queue package, this review will be closed. A fresh PR for fresh review will be started when ready, and comments leftover from this review will still be addressed.

jeffhli commented:
> With the vast internal design changes through this PR's history, alongside new work to support the queue package, this review will be closed. A fresh PR for fresh review will be started when ready, and comments leftover from this review will still be addressed.

Does this PR have a milestone?

immarvin commented:
hi @jaschrep-msft , is the new PR you mentioned created? any planned date on the release of client side encryption feature? thx
